### PR TITLE
feat: refactor to allow unlimited facets - WIP

### PR DESCRIPTION
**Pull request by stephanegigandet**
Some initial refactoring to allow to have an unlimited number of <tag type>/<tag> in URLs.

image

### CODE REVIEW

File: lib/ProductOpener/Display.pm [hunk 1]
```diff
@@ -3033,81 +3033,70 @@ sub display_tag ($request_ref) {
...
+	foreach my $tag_ref (@{$request_ref->{tags}}) {
 
-		$header .= '<meta name="robots" content="noindex">' . "\n";
-
-	}
+		my $tagid = $tag_ref->{tagid};
+		my $tagtype = $tag_ref->{tagtype};
+		my $tag_prefix = $tag_ref->{tag_prefix};
+		my $display_tag;
+		my $canon_tagid;
```
Comment (alexgarel) [inline lib/ProductOpener/Display.pm:3054]:
```suggestion
		# canonical tag corresponding to tagid
		my $canon_tagid;
```

File: lib/ProductOpener/Display.pm [hunk 2]
```diff
@@ -3033,81 +3033,70 @@ sub display_tag ($request_ref) {
...
-		)
-	{
+	foreach my $tag_ref (@{$request_ref->{tags}}) {
 
-		$header .= '<meta name="robots" content="noindex">' . "\n";
-
-	}
+		my $tagid = $tag_ref->{tagid};
+		my $tagtype = $tag_ref->{tagtype};
+		my $tag_prefix = $tag_ref->{tag_prefix};
```
Comment (alexgarel) [inline lib/ProductOpener/Display.pm:3052]: can you add a comment on what is tag_prefix ?

File: lib/ProductOpener/Display.pm [hunk 3]
```diff
@@ -3033,81 +3033,70 @@ sub display_tag ($request_ref) {
...
-		or (    (defined $tagtype2)
-			and ($tagtype2 =~ /^(users|correctors|editors|informers|correctors|photographers|checkers)$/))
-		)
-	{
+	foreach my $tag_ref (@{$request_ref->{tags}}) {
 
-		$header .= '<meta name="robots" content="noindex">' . "\n";
-
-	}
+		my $tagid = $tag_ref->{tagid};
```
Comment (alexgarel) [inline lib/ProductOpener/Display.pm:3050]:
```suggestion
		# the tag name requested in url (in $lc language)
		my $tagid = $tag_ref->{tagid};
```

File: lib/ProductOpener/Display.pm [hunk 4]
```diff
@@ -3033,81 +3033,70 @@ sub display_tag ($request_ref) {
...
+		my $tagtype = $tag_ref->{tagtype};
+		my $tag_prefix = $tag_ref->{tag_prefix};
+		my $display_tag;
+		my $canon_tagid;
+		my $newtagid;
```
Comment (alexgarel) [inline lib/ProductOpener/Display.pm:3055]:
```suggestion
		# tag name in $lc corresponding to $canon_tagid
		# (if it's not the same as $tagid, we will redirect the request)
		my $newtagid;
```

File: lib/ProductOpener/Routing.pm [hunk 5]
```diff
@@ -408,152 +408,116 @@ sub analyze_request ($request_ref) {
 				if $log->is_debug();
 		}
 
-		if (
+		# Extract tag type / tag value pairs and store them in an array $request_ref->{tags}
```
Comment (alexgarel) [inline lib/ProductOpener/Routing.pm:411]: Could we extract this part in a separate sub (making code easier to read) ?

File: lib/ProductOpener/Test.pm [hunk 6]
```diff
@@ -743,6 +744,24 @@ sub normalize_products_for_test_comparison ($array_ref) {
...
+
+=head3 Arguments
+
+=head4 array_ref
+
+Array of products
+
+=cut
+
+sub sort_products_for_test_comparison ($array_ref, $sort_field) {
```
Comment (alexgarel) [inline lib/ProductOpener/Test.pm:759]: Good idea to make it a sub even if it's a one-liner, it really helps understand the code more easily.

File: tests/integration/facets.t [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+		expected_status_code => 200,
+		sort_products_by => 'product_name',
+	},
+];
+
+# note: we need to execute the tests with bob, because we need authentication
+# to see data quality panels
+execute_api_tests(__FILE__, $tests_ref, $ua);
+
+done_testing();
```
Comment (alexgarel) [inline tests/integration/facets.t:204]: :heart_eyes: some more integration tests

**Review by alexgarel — APPROVED** [review]
Great move !

I propose a small refactoring to have even better readability.

File: lib/ProductOpener/Display.pm (hunk 2 above)
Comment (stephanegigandet) [reply lib/ProductOpener/Display.pm:3000]: sure

File: lib/ProductOpener/Routing.pm (hunk 5 above)
Comment (stephanegigandet) [reply lib/ProductOpener/Routing.pm:411]: Sure.

**Pull request merged**